Counterexamples/HarmonicNotInteger.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2023 Koundinya Vajjha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Koundinya Vajjha
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.Data.Int.Log
+/-!
+
+The nth Harmonic number is not an integer. We formalize the proof using
+2-adic valuations. This proof is due to Kürschák.
+
+Reference:
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/padicharmonicsum.pdf
+
+-/
+
+namespace Counterexample
+
+
+open BigOperators
+
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, 1 / (i + 1)
+
+lemma harmonic_pos (n : ℕ) (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => div_pos zero_lt_one
+  (by norm_cast; linarith))
+  (by rwa [Finset.nonempty_range_iff])
+
+lemma harmonic_singleton {n c : ℕ} (hc : c ∈ Finset.range n):
+    harmonic n = 1 / (c + 1) + ∑ x in Finset.range n \ {c}, 1 / ((x : ℚ) + 1) := by
+  unfold harmonic
+  rwa [add_comm, Finset.sum_eq_sum_diff_singleton_add (i := c)]
+
+lemma finset_range_sdiff_singleton_nonempty {c n : ℕ} (hn : 2 ≤ n):
+    Finset.Nonempty (Finset.range n \ {c}) := by
+  rw [Finset.sdiff_nonempty, Finset.subset_singleton_iff, Finset.range_eq_empty_iff, not_or]
+  refine' ⟨by linarith, fun Hnot => _⟩
+  suffices n = 1 by linarith
+  rw [← Finset.card_range n, ← Finset.card_singleton c, Hnot]
+
+
+lemma harmonic_singleton_ne_zero {c n : ℕ} (hn : 2 ≤ n) :
+    ∑ x in Finset.range n \ {c}, 1 / (x + 1 : ℚ) ≠ 0 := by
+  refine' ne_of_gt <| Finset.sum_pos (fun i _ => div_pos zero_lt_one _)
+    (finset_range_sdiff_singleton_nonempty hn)
+  norm_cast; simp only [add_pos_iff, or_true]
+
+/-- If a rational is not a 2-adic integer, it is not an integer. -/
+theorem not_int_of_not_padic_int {a : ℚ} (H : 1 < padicNorm 2 a):
+    ¬ a.isInt := by
+  suffices : a.den ≠ 1; simpa [Rat.isInt]
+  by_cases h : a = 0
+  · subst h; contradiction
+  · unfold padicNorm at H
+    split_ifs at H; contradiction
+    suffices padicValRat 2 a < 0 by
+      unfold padicValRat at this
+      intro Hden
+      rw [Hden] at this; simp only [padicValNat.one, sub_zero] at this
+      norm_cast
+    refine' neg_of_neg_pos _
+    have hx : (1 : ℚ) < 2 := by norm_num
+    rw [← zpow_lt_iff_lt hx]
+    simpa only [zpow_zero]
+
+
+lemma padicValRat_2_pow (r : ℕ) : padicValRat 2 (1 / 2^r) = -r := by
```

Comment:
we tend not to use numbers in names but instead spell them out

Reply:
That said surely this lemma has nothing to do with two specificly? and could be generalized to any p?

Reply:
yes but i really only needed 2-adic valuations in this proof, but this was more of a local lemma specific to this file which i needed in the proofs below.

Reply:
(but i can generalize it)

Reply:
Mathlib philosophy is really to try and avoid local lemmas that are only useful once, especially when the generalization is easy and the lemma could be useful to someone else at some point

Reply:
yes, the lemma is generalized in 38884190410085716619b1cf03d126b033546b90.

Reply:
Yes, that lemma is generalized in 38884190410085716619b1cf03d126b033546b90.

Reply:
Yeah I was just trying to explain a bit as you sounded unconvinced of the merits of generalizing!